src/cogent3/app/composable.py

```diff
@@ -784,6 +787,29 @@ class AppType(Enum):
 GENERIC = AppType.GENERIC
 
 
+def _get_raw_hints(main_func, min_params):
+    _no_value = inspect.Parameter.empty
+    params = inspect.signature(main_func)
+    if len(params.parameters) < min_params:
+        raise ValueError(
+            f"{main_func.__name__!r} must have at least {min_params} input parameters"
+        )
+    # annotation for first parameter other than self, params.parameters is an orderedDict
+    first_param_type = [p.annotation for p in params.parameters.values()][1]
```

Comment:
If we can have `min_params > 1`, then I think this index should be: 
`[1:min_params]`
and return frozenset of hints

Reply:
I disagree. This first provided type hint is used to dictate compatibility for composition because it's the primary input.

For the decorated class case, there must be 2 parameters: `main(self, data)`. But for the decorated function I'm now thinking we should just allow it to be one.